cpp/src/arrow/compute/kernels/scalar_compare_test.cc

```diff
@@ -1850,5 +1851,154 @@ TEST(TestMaxElementWiseMinElementWise, CommonTemporal) {
+  checkTimestampArray(timestamp(unit), "[]", l, r, "[]");
+  checkTimestampArray(timestamp(unit), "[null]", l, r, "[null]");
+  checkTimestampArray(timestamp(unit), R"(["1970-01-01","2000-02-29","1900-02-28"])", l,
+                      r, "[false,true,false]");
+  checkTimestampArray(timestamp(unit), R"(["1970-01-01","2000-02-29","2004-02-28"])", l,
+                      r, "[false,true,true]");
+  checkTimestampArray(timestamp(unit), R"(["2018-01-01","1999-04-04","1900-02-28"])", l,
+                      r, "[true,false,false]");
+}
+
+TYPED_TEST(TestNumericBetweenKernel, SimpleBetweenArrayArrayArray) {
+  ValidateBetween<TypeParam>(
+      "[]", ArrayFromJSON(TypeTraits<TypeParam>::type_singleton(), "[]"),
+      ArrayFromJSON(TypeTraits<TypeParam>::type_singleton(), "[]"), "[]");
+  ValidateBetween<TypeParam>(
+      "[null]", ArrayFromJSON(TypeTraits<TypeParam>::type_singleton(), "[null]"),
+      ArrayFromJSON(TypeTraits<TypeParam>::type_singleton(), "[null]"), "[null]");
+  ValidateBetween<TypeParam>(
+      "[1,1,2,2,2]",
+      ArrayFromJSON(TypeTraits<TypeParam>::type_singleton(), "[0,0,1,3,3]"),
+      ArrayFromJSON(TypeTraits<TypeParam>::type_singleton(), "[10,10,2,5,5]"),
+      "[true,true,false,false,false]");
+  ValidateBetween<TypeParam>(
+      "[1,1,2,2,2,2]",
+      ArrayFromJSON(TypeTraits<TypeParam>::type_singleton(), "[0,0,1,null,3,3]"),
+      ArrayFromJSON(TypeTraits<TypeParam>::type_singleton(), "[10,10,2,2,5,5]"),
+      "[true,true,false,null,false,false]");
+}
+
+TEST(TestSimpleBetweenKernel, StringArrayArrayArrayTest) {
+  ValidateBetween<StringType>(
+      R"(["david","hello","world"])",
+      ArrayFromJSON(TypeTraits<StringType>::type_singleton(), R"(["adam","hi","whirl"])"),
+      ArrayFromJSON(TypeTraits<StringType>::type_singleton(),
+                    R"(["robert","goeiemoreen","whirlwind"])"),
+      "[true, false, false]");
+  ValidateBetween<StringType>(
+      R"(["x","a","f"])",
+      ArrayFromJSON(TypeTraits<StringType>::type_singleton(), R"(["w","a","e"])"),
+      ArrayFromJSON(TypeTraits<StringType>::type_singleton(), R"(["z","a","g"])"),
+      "[true, false, true]");
+  ValidateBetween<StringType>(
+      R"(["block","bit","binary"])",
+      ArrayFromJSON(TypeTraits<StringType>::type_singleton(),
+                    R"(["bit","nibble","ternary"])"),
+      ArrayFromJSON(TypeTraits<StringType>::type_singleton(), R"(["word","d","xyz"])"),
+      "[true, false, false]");
+  ValidateBetween<StringType>(R"(["Ayumi","アユミ","王梦莹"])",
+                              ArrayFromJSON(TypeTraits<StringType>::type_singleton(),
+                                            R"(["たなか","あゆみ","歩美"])"),
+                              ArrayFromJSON(TypeTraits<StringType>::type_singleton(),
+                                            R"(["李平之","田中","たなか"])"),
+                              "[false, true, false]");
+}
+
+TEST(TestSimpleBetweenKernel, TimestampArrayArrayArrayTest) {
```

Comment:
Do we also need different time unit case?

Reply:
Will add time tests

Reply:
@kou @jorisvandenbossche Raised an issue to use a tool to measure testing coverage https://issues.apache.org/jira/browse/ARROW-15308 - some of the Python and other language binding tests will duplicate C++ tests. Possibly some test rationalization is required since one wants to also minimize testing time.

Reply:
Added different time units.